### ISSUE TITLE
Keras Accuracy with sequence auto-encoder is different than manual calculation

### ISSUE THREAD

**Issue by brijesh-6899** [tensorflow/tensorflow#42304]
Hi,

I'm using GRU-based seq2seq model with byte pair encoding for spelling correction problem. 
_This is my model architecture/code:_

```
encoder_inputs = Input(shape=(padding_length,), name="EncoderInput_1")
embedded_encoder_inputs = Embedding(num_encoder_tokens, latent_dim, mask_zero=True)(encoder_inputs)
encoder = GRU(latent_dim, return_state=True)
_, state_h = encoder(embedded_encoder_inputs)

encoder_states = state_h

decoder_inputs = Input(shape=(padding_length,), name="DecoderInput_1")
embedded_decoder_inputs = Embedding(num_decoder_tokens, latent_dim, mask_zero=True)(decoder_inputs)
decoder_lstm = GRU(latent_dim, return_sequences=True, return_state=True)
x, _ = decoder_lstm(embedded_decoder_inputs, initial_state=encoder_states)
decoder_dense = TimeDistributed(Dense(num_decoder_tokens, activation='softmax'))
decoder_outputs = decoder_dense(x)

model = Model([encoder_inputs, decoder_inputs], decoder_outputs)

model.compile(optimizer='adam', loss=sparse_categorical_crossentropy,metrics=["accuracy"])
model.fit([ip_encoded,op_encoded],
              op_offset_encoded,
              epochs=epochs,
              verbose=1,
              batch_size=batch_size,
              validation_split=0.2,
```
I trained it for 10 epochs with ~700K training examples and the final validation accuracy was ~97%.

I looked into the sample predictions and accuracies (I am using model.evaluate to get single example accuracy), but it is deviating from manual calculations. 

I'd like to mention my understanding of the accuracy calculation which I'm using during manual calculation:
- For all Input tokens to decoder, it predicts the next token. If predicted token == expected token, the "count" of correctly predicted is increased by 1. I apply the same for all decoder Input tokens and at the last calculate accuracy by dividing "count" with the count of total Input tokens.

Examples: ( ip: input to model, op: model output, ex: expected)
_(Note: I'm assuming padding tokens are not considered for metric calculation)_
_(Note: \<start\>: 1, \<pad\>: 0, \<end\>: 2)_

Code used for finding accuracy for single example:
`model.evaluate([ip_encoded[idx:idx+1],op_encoded[idx:idx+1]],op_offset_encoded[idx:idx+1],batch_size=64)`

1. (all tokens are predicted correctly, manual and Keras accuracies are same)
```
ip:  Do you get a pattern similar to t hat shown in Fig.
op:  Do you get a pattern similar to that shown in Fig.
ex:  Do you get a pattern similar to that shown in Fig.

ip:  [1, 1313, 416, 742, 261, 1953, 1239, 299, 259, 308, 270, 911, 283, 814, 16, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
op:  [1313, 416, 742, 261, 1953, 1239, 299, 336, 911, 283, 814, 16, 2]
ex:  [1313, 416, 742, 261, 1953, 1239, 299, 336, 911, 283, 814, 16, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

1/1 [==============================] - 0s 1ms/step - loss: 0.0031 - accuracy: 1.0000
[0.003069164464250207, 1.0]
```
2. (manual accuracy: 17/23=0.73, keras: 0.95)
```
ip:  The equal part that we get in is, therefore, larger than the equal part we get in.
op:  The equal part that we get in  is, therefore, larger than the equal part in which we get .
ex:  The equal part that we get in  is, therefore, larger than the equal part we get in .

ip:  [1, 316, 752, 496, 336, 370, 742, 283, 294, 14, 1578, 14, 2115, 621, 264, 752, 496, 370, 742, 283, 16, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
op:  [316, 752, 496, 336, 370, 742, 283, 223, 294, 14, 1578, 14, 2115, 621, 264, 752, 496, 283, 406, 370, 742, 1964, 2]
ex:  [316, 752, 496, 336, 370, 742, 283, 223, 294, 14, 1578, 14, 2115, 621, 264, 752, 496, 370, 742, 283, 1964, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

1/1 [==============================] - 0s 1ms/step - loss: 0.0289 - accuracy: 0.9565
[0.028948727995157242, 0.95652174949646]
```
3. (manual: 10/13=0.77, Keras: 0.92)
```
ip:  Can you tell what advanxtage the fungus derives from this association?
op:  Can you tell what advantage affect the sewage from this association?
ex:  Can you tell what advantage the fungus derives from this association?

ip:  [1, 1321, 416, 2797, 985, 2923, 282, 1424, 596, 264, 9590, 19016, 389, 428, 6330, 33, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
op:  [1321, 416, 2797, 985, 6089, 2211, 264, 4739, 389, 428, 6330, 33, 2]
ex:  [1321, 416, 2797, 985, 6089, 264, 9590, 19016, 389, 428, 6330, 33, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

1/1 [==============================] - 0s 1ms/step - loss: 0.0121 - accuracy: 0.9286
[0.012137999758124352, 0.9285714030265808]
```

- It'd be really helpful if community can help me understand Keras accuracy metrics internals and verify my manual metric calculation process.

**Comment by Saduf2019** [comment]
@brijesh-6899 
This question is better asked on [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.
